keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1243 @@
+Linkerd and Istio), need to do several hacks to have the basic functionality.
+
+Some service meshes depend on secret (like a certificate) downloaded by other
+Init container to establish secure communication between services. This makes
+the problem of ordering of sidecar and Init containers harder.
+
+### Problems: configuration / secrets
+
+Some pods use init containers to pull down configuration/secrets and update them
+before the main container gets it. Then use sidecars to continue to watch for
+changes and perform the updates and push to the main container. This requires
+two separate code paths today. Perhaps the same sidecar could be used for both
+cases.
+
+### Goals
+
+This proposal aims to:
+- make containers implementing the sidecar pattern first class citizens inside a
+  Pod
+- solve a Job completion issue when sidecars should run continuously
+- allow mixing initContainers and sidecars for a choreographed startup sequence
+- allow easy injection of sidecar containers in any Pod
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- act as a security control to enforce that pod containers only run while the
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- allow to enforce security boundaries to sidecar containers different that
+  other containers. For example, allow Istio to run privileged to configure ip
+  tables and disable other containers from doing so
+- (alpha) support containers termination ordering
+
+## Proposal
+
+The proposal is to introduce a `restartPolicy` field to Init containers and use
+it to indicate that an initContainer is a sidecar container. The only supported
+value proposed in this KEP is `Always`. Kubelet will start these containers in
+the order with other Init containers, but instead of waiting for its completion,
+it will wait for the container startup completion (startup probe succeeded or if
+no startup probe defined, `postStart` handler completed).
+
+Other values for `restartPolicy` of containers will not be accepted and
+containers will follow the logic is currently implemented (documented
+[here](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#:~:text=if%20a%20pod's%20init%20container%20fails%2C%20the%20kubelet%20repeatedly%20restarts%20that%20init%20container%20until%20it%20succeeds.%20however%2C%20if%20the%20pod%20has%20a%20restartpolicy%20of%20never%2C%20and%20an%20init%20container%20fails%20during%20startup%20of%20that%20pod%2C%20kubernetes%20treats%20the%20overall%20pod%20as%20failed.)
+and more details can be found in the next section "Future use of `restartPolicy`
+field").
+
+Sidecar containers will not block Pod completion - if all regular containers
+complete, sidecar containers will be terminated.
+
+During the sidecar startup stage it will behave similarly to Init containers. If
+Pod `restartPolicy` is `Never`, sidecar container failed during startup will NOT
```

Comment:
If restartPolicy is Never it is not a side car, right?

Reply:
This refers to Pod `restartPolicy`